### PR TITLE
Document the design of Async#cont

### PR DESCRIPTION
**Pull request by TimWSpence**

### CODE REVIEW

File: docs/typeclasses/async.md [hunk 1]
```diff
@@ -102,3 +102,118 @@ for {
...
+
+We could then define `async` in terms of `cont`
+
+```scala
+def async[A](k: (Either[Throwable, A] => Unit) => F[Option[F[Unit]]]): F[A] =
+  cont flatMap {
+    case (cb, fa) => 
+      k(cb).flatMap(_.map(fa.onCancel(_)).getOrElse(fa))
+  }
+```
```
Comment (SystemFw) [inline docs/typeclasses/async.md:184]: you probably want to introduce `uncancelable` here first

File: docs/typeclasses/async.md [hunk 2]
```diff
@@ -102,3 +102,118 @@ for {
...
+
+In this case, the error channel is propagated from `k(cb)` as we directly `flatMap`
+on the result.
+
+Fantastic! We're done, right? Well... not quite. The problem is that it is not safe
+to call concurrent operations such as `get.start`. We therefore need to employ
```
Comment (SystemFw) [inline docs/typeclasses/async.md:190]: yeah this is hard to explain, but essentially the reason is that `get` represents "the rest of the runloop", which don't want to be resumed concurrently from multiple places

File: docs/typeclasses/async.md (hunk 1 above)
Comment (TimWSpence) [reply docs/typeclasses/async.md:184]: Yep thanks 👍 Just copied this from Gitter without thinking too much about it 😂 Although I did also consider removing `uncancelable` from both implementations as it's pretty irrelevant to what we're actually discussing here (could leave a note to say that we're ignoring cancellation). What do you think?

File: docs/typeclasses/async.md [hunk 3]
```diff
@@ -102,3 +102,123 @@ for {
...
+## Here be dragons
+
+`Async` also defines a function called `cont`. This is _not_ intended to be
+called from user code. You also _absolutely do not_ need to understand it in
+order to use `Async`. The design is however very instructive for the curious
+reader. To start with, consider the definition of `async` again
+
+```scala
+trait Async[F[_]] {
+  def async[A](k: (Either[Throwable, A] => Unit) => F[Option[F[Unit]]]): F[A] = {
```
Comment (djspiewak) [inline docs/typeclasses/async.md]: Random open curly? :-)

File: docs/typeclasses/async.md [hunk 4]
```diff
@@ -102,3 +102,123 @@ for {
...
+This looks (vaguely) reasonable. However, there is a subtle problem lurking in
+there -  the use of `flatten`. The problem is that `OptionT` (and similarly
+`EitherT` and `IorT`) has an additional error channel (where `x = 
+delegrate.pure(None)`) that we are forced to discard when mapping to something
+of type `F[Option[F[Unit]]]`.
+
+This in fact breaks several of the `Async` laws:
+
+```scala
+def asyncRightIsSequencedPure[A](a: A, fu: F[Unit]) =
```
Comment (djspiewak) [inline docs/typeclasses/async.md:147]: Needs indentation

File: docs/typeclasses/async.md [hunk 5]
```diff
@@ -102,3 +102,123 @@ for {
...
+    F.async[A](k => F.delay(k(Right(a))) >> fu.as(None)) <-> (fu >> F.pure(a))
+
+  def asyncLeftIsSequencedRaiseError[A](e: Throwable, fu: F[Unit]) =
+    F.async[A](k => F.delay(k(Left(e))) >> fu.as(None)) <-> (fu >> F.raiseError(e))
+```
+
+In both cases if we have `fu = OptionT.none[F, A]` (the error channel for `OptionT`) then
+the LHS will suppress the error, whereas the RHS will not.
+
+A possible solution would be to un-CPS the computation by defining
```
Comment (djspiewak) [inline docs/typeclasses/async.md]: Maybe link `CPS` to the [Wikipedia description](https://en.wikipedia.org/wiki/Continuation-passing_style)?

**Review by djspiewak — CHANGES_REQUESTED** [review]
Looks really really nice! A couple changes but otherwise this is awesome

File: docs/typeclasses/async.md [hunk 6]
```diff
@@ -102,3 +102,123 @@ for {
...
+      }
+    }
+  )
+}
+```
+
+This looks (vaguely) reasonable. However, there is a subtle problem lurking in
+there -  the use of `flatten`. The problem is that `OptionT` (and similarly
+`EitherT` and `IorT`) has an additional error channel (where `x = 
+delegrate.pure(None)`) that we are forced to discard when mapping to something
```
Comment (catostrophe) [inline docs/typeclasses/async.md]: ~delegrate~ delegate

File: docs/typeclasses/async.md (hunk 3 above)
Comment (TimWSpence) [reply docs/typeclasses/async.md]: 🤦 thanks!

**Comment by TimWSpence** [comment]
@djspiewak I _believe_ I've addressed all your comments. Apologies if I've missed any!

**Pull request merged**